### PR TITLE
Feature: navigation behavioural tests [NP-1751]

### PR DESCRIPTION
**Pull request by davidsauntson**
Baby's first cucumber.  Please go wild!

Wanted to add some behavioural tests for Greedy Nav as:
 - there aren't any
 - I want to practice
 - we are about to refactor it and the current tests rely on testing internals
 - the refactor involves moving to `IntersectionObserver` which isn't really testable in a JSDOM unit test, so behavioural tests seem appropriate

### CODE REVIEW

File: testing/features/navigation.feature
Comment (davidsauntson) [inline testing/features/navigation.feature:35]: This currently works on live because of the presence of header links

File: testing/features/support/components/navigation/default.rb
Comment (davidsauntson) [inline testing/features/support/components/navigation/default.rb:7]: I've tried to go for a more user based approach (eg find buttons by text), which means this class is basically empty.  Is that...ok?

File: testing/features/support/driver.rb
Comment (davidsauntson) [inline testing/features/support/driver.rb:28]: Not sure if we would want this to be 0?  I'm aware I've only used capybara selectors so not sure if there is a SitePrism thing I should be using?  Or if this should just be set for the nav tests (the delay is required to wait for the More and Close buttons to be present).

File: testing/features/step_definitions/navigation_steps.rb
Comment (davidsauntson) [inline testing/features/step_definitions/navigation_steps.rb:57]: feels a bit brittle

File: testing/features/navigation.feature
Comment (luke-hill) [reply testing/features/navigation.feature:35]: Rightio. Is this something we have a ticket for. We could add it to this as an `@` tag like `@NP-1244`

File: testing/features/step_definitions/navigation_steps.rb
Comment (luke-hill) [inline testing/features/step_definitions/navigation_steps.rb:11]: We have a tag which will do this for you. I believe it is something like `@requires_small_screen` or something similar.

File: testing/features/step_definitions/navigation_steps.rb
Comment (luke-hill) [inline testing/features/step_definitions/navigation_steps.rb:15]: Based on the way you have written this cuke (Gherkin step). This is really good. As it's more imperative. I'll put some suggestions to keep your other ones like this.

File: testing/features/step_definitions/navigation_steps.rb
Comment (luke-hill) [inline testing/features/step_definitions/navigation_steps.rb:28]: For this one. Do we want to validate that it has been set to invisible? Or do we not care? (Just noticed you've switched from not present to hidden)

File: testing/features/step_definitions/navigation_steps.rb
Comment (luke-hill) [inline testing/features/step_definitions/navigation_steps.rb:44]: `the dropdown menu has header links`

File: testing/features/step_definitions/navigation_steps.rb
Comment (luke-hill) [inline testing/features/step_definitions/navigation_steps.rb:45]: For the signature here you can chain things. So you can do something like

`expect(@component.dropdown_menu).to have_header_links`

File: testing/features/step_definitions/navigation_steps.rb
Comment (luke-hill) [reply testing/features/step_definitions/navigation_steps.rb:57]: We have a tab helper. It's in the header tests (Which caters to safari which is a bit odd).

It might not work directly out of the box. But if you can try "hook" into that. Maybe by providing an additional keyword arg or something - I'll let you run wild.

File: testing/features/support/components/navigation/default.rb
Comment (luke-hill) [reply testing/features/support/components/navigation/default.rb:7]: So if you need to directly use text. I'm not 100% for it, but if enough people decide to then I guess I'll have to cede.

2 things to note.
1) Do these text options change by language? If so should we be doing it?
2) We can still create POM elements using capybara locator args.

Essentially the declaration from item 3 onwards (The hashable args). Is a capybara locator strategy
`element` is the keyword method definition
`:symbol`, is the reference name
`PascalCaseClass` is the class (Only used in Section/Sections)
`{hash: args}` are then the arguments to locate. So here you could do something like `"button", text: "Hey there!"`

File: testing/features/support/driver.rb
Comment (luke-hill) [reply testing/features/support/driver.rb:28]: So this is a setting we should be very careful with editing. Ideally we would use in-line waiters. Any declaration or invocation of a CNE (Capybara::Node::Element), either in-line with capybara or using SitePrism, can pass it.

i.e.

```
page.find("button", wait: 5) # Wait up to 5secs for button
@component.navigation(wait: 3) # Wait up to 3secs for navigation SitePrism element
element :navigation, ".js-cads-greedy-nav", wait: 3 # Wait up to 3 secs to "set" the SitePrism reference
```

If you find you need to set an implicit waiting strategy, and you're using something that isn't Mojave Safari, I'd be very intrigued why it's needed.

**Review by luke-hill — COMMENTED** [review]
I've tried to not focus on style / grammar for now. As that again is a subjective view.

We tend to have a few more "Behavioural" style definitions where possible. I appreciate in the DS it's crap, because we don't really have proper user stories, and it's more of an integration environment. I can maybe try get a feel for how you want to tweak things or pair on how I come to certain vocab style decisions?

File: testing/features/step_definitions/navigation_steps.rb
Comment (davidsauntson) [reply testing/features/step_definitions/navigation_steps.rb:28]: Yeah I guess it should be `not present` for consistency 👍

File: testing/features/support/components/navigation/default.rb
Comment (davidsauntson) [reply testing/features/support/components/navigation/default.rb:7]: Agreed to move to using the SitePrism elements

File: testing/features/support/driver.rb
Comment (davidsauntson) [reply testing/features/support/driver.rb:28]: Agreed to use inline waiting where required

File: testing/features/step_definitions/navigation_steps.rb
Comment (davidsauntson) [reply testing/features/step_definitions/navigation_steps.rb:57]: Will potentially look at refactoring as it's tightly coupled to testing the skip links currently.

File: testing/features/step_definitions/navigation_steps.rb
Comment (luke-hill) [inline testing/features/step_definitions/navigation_steps.rb:11]: Everywhere this and the other resize logic is used. If you just use a tag of `@small_screen` it will work for both mobiles and non mobiles, and saves you a load of steps.

**Review by luke-hill — COMMENTED** [review]
Looking really good. Couple of minor tidy-ups but approach seems fine

File: testing/features/step_definitions/navigation_steps.rb
Comment (davidsauntson) [reply testing/features/step_definitions/navigation_steps.rb:11]: @luke-hill Not sure about implementing this.  It litters the gherkin with `@small_screen` to remove a single step definition.  The `Given` clauses will obviously need to stay in the gherkin as it is part of the scenario.  Using  the `@small_screen` tag with `Background:` instead of `Scenario:` produces an obscure error: 

```
When using the Selenium Logger that is set to pipe to a file, the Net::HTTP adapter (default),
can return unencoded binary (confusingly called ASCII-8BIT in Ruby).
Consequently we set the logger to binmode so it doesn't try to encode the data - this would always
cause errors for non-ASCII characters, whatever the parent encoding is. An example of this is ©.
```

File: testing/features/support/components/navigation/standard.rb
Comment (davidsauntson) [reply testing/features/support/components/navigation/standard.rb:12]: OK, but this is the opposite of what we discussed yesterday.

Comment (davidsauntson) [reply testing/features/support/components/navigation/standard.rb]: Would be useful to get some rationale / guidance on what you expect in a step definition file and what you expect to be in the page object model and, importantly, why.

File: testing/features/step_definitions/navigation_steps.rb
Comment (luke-hill) [reply testing/features/step_definitions/navigation_steps.rb:11]: Fair enough. i think this is one we can table to fix later.

Ordinarily I'd tell you to tag the rule - Which then saves you a whole heap of lines - However you can't do this (Yet).

The only issue with this is this logic will fail on mobiles. So if you check in `hooks.rb` I'm happy if you want to copy the conditional there, and I can make a tech debt ticket to refactor this feature later on once we've done the upgrades required.

Comment (davidsauntson) [reply testing/features/step_definitions/navigation_steps.rb:11]: I don't know what this means.  What logic? Resizing the screen?  Are you saying I need to tag these with `@not_mobile`?

Comment (luke-hill) [reply testing/features/step_definitions/navigation_steps.rb:11]: Ah sorry. No not quite.

So if you see here: https://github.com/citizensadvice/design-system/blob/master/testing/features/support/hooks.rb#L12-L13 - You'll see the logic that would be desirable.

But for now, as a short-term fix. Just whacking a `unless mobile_phone?` at the end of your resizing logic should be fine. I'll tidy everything up down the line once the requisite updates are done.

Comment (davidsauntson) [reply testing/features/step_definitions/navigation_steps.rb:11]: Great, thanks.

**Review by luke-hill — APPROVED** [review]

File: testing/features/step_definitions/navigation_steps.rb
Comment (pedantic-git) [inline testing/features/step_definitions/navigation_steps.rb:4]: General question more for @luke-hill here - how come this is the conventional way to call `#load` on these classes? Is there scope for making it either one of these three things:

1. Happen automatically as part of `.new` (I'm guessing there are good reasons not to do this one!)
2. Happen if a given parameter is passed to `.new` (e.g. `Navigation::Standard.new(load: true)`)
3. Or at least have the `#load` method return `self` so that you can call it as `.new.load` instead of `.new.tap(&:load)`?

**Comment by pedantic-git** [comment]
This was a really good read! Sorry I didn't get to it before it merged but have a stamp of approval anyway!

File: testing/features/step_definitions/navigation_steps.rb
Comment (luke-hill) [reply testing/features/step_definitions/navigation_steps.rb:4]: So.

`#load` as a method will do about 4 things under the hood: https://github.com/site-prism/site_prism/blob/main/lib/site_prism/page.rb#L39-L57

When you instantiate a new page class, it just returns an OO object. So you can't do anything on instantiation (Yet, see below).
When instantiating a page object. We don't like users passing in much (As a few things need to be passed in).
The load method "could" return `self`, but then that conflicts with the specification, which says it returns a yielded object (This is to support Capybara simple / string objects).
By default most of the time load will return `true`. This is to stipulate that it has passed the requisite load validation tests.

However, that all being said, a long standing branch of mine: https://github.com/site-prism/site_prism/tree/feature/auto_run_validations would allow load validations to be automatically ran when calling `#new` - But I've not got time / priority to work on this. This is something that will likely land in 4.x at some point.

Comment (pedantic-git) [reply testing/features/step_definitions/navigation_steps.rb:4]: @luke-hill This is very clear. Thank you for the explanation!

**Pull request merged**
